src/main/java/com/yahoo/memory/NonNativeWritableMemoryImpl.java

```diff
@@ -0,0 +1,306 @@
+/*
+ * Copyright 2017, Yahoo! Inc. Licensed under the terms of the
+ * Apache License 2.0. See LICENSE file at the project root for terms.
+ */
+
+package com.yahoo.memory;
+
+import static com.yahoo.memory.UnsafeUtil.ARRAY_DOUBLE_INDEX_SCALE;
+import static com.yahoo.memory.UnsafeUtil.ARRAY_FLOAT_INDEX_SCALE;
+import static com.yahoo.memory.UnsafeUtil.ARRAY_LONG_INDEX_SCALE;
+import static com.yahoo.memory.UnsafeUtil.CHAR_SHIFT;
+import static com.yahoo.memory.UnsafeUtil.DOUBLE_SHIFT;
+import static com.yahoo.memory.UnsafeUtil.FLOAT_SHIFT;
+import static com.yahoo.memory.UnsafeUtil.INT_SHIFT;
+import static com.yahoo.memory.UnsafeUtil.LONG_SHIFT;
+import static com.yahoo.memory.UnsafeUtil.SHORT_SHIFT;
+import static com.yahoo.memory.UnsafeUtil.unsafe;
+
+import java.nio.ByteOrder;
+
+/*
+ * Developer notes: The heavier methods, such as put/get arrays, duplicate, region, clear, fill,
+ * compareTo, etc., use hard checks (checkValid*() and checkBounds()), which execute at runtime and
+ * throw exceptions if violated. The cost of the runtime checks are minor compared to the rest of
+ * the work these methods are doing.
+ *
+ * <p>The light weight methods, such as put/get primitives, use asserts (assertValid*()), which only
+ * execute when asserts are enabled and JIT will remove them entirely from production runtime code.
+ * The light weight methods will simplify to a single unsafe call, which is further simplified by
+ * JIT to an intrinsic that is often a single CPU instruction.
+ */
+
+/**
+ * Implementation of {@link WritableMemory} for non-native endian byte order. Native variant is
+ * {@link WritableMemoryImpl}.
+ * @author Roman Leventov
+ * @author Lee Rhodes
+ */
+final class NonNativeWritableMemoryImpl extends BaseWritableMemoryImpl {
+
+  NonNativeWritableMemoryImpl(final ResourceState state, final boolean localReadOnly) {
+    super(state, localReadOnly);
+    if (state.getResourceOrder() == ByteOrder.nativeOrder()) {
+      throw new IllegalStateException(
+          "Expected non-native ordered state. This should be a bug in the Memory library.");
+    }
+  }
+
+  //REGIONS XXX
+  @Override
+  public Memory region(final long offsetBytes, final long capacityBytes) {
+    return writableRegionImpl(offsetBytes, capacityBytes, true);
+  }
+
+  @Override
+  public WritableMemory writableRegion(final long offsetBytes, final long capacityBytes) {
+    return writableRegionImpl(offsetBytes, capacityBytes, localReadOnly);
+  }
+
+  private WritableMemory writableRegionImpl(final long offsetBytes, final long capacityBytes,
+      final boolean localReadOnly) {
+    checkValidAndBounds(offsetBytes, capacityBytes);
+    if (capacityBytes == 0) { return WritableMemoryImpl.ZERO_SIZE_MEMORY; }
```

Comment:
The static field BaseWritableMemoryImpl.ZERO_SIZE_MEMORY should be accessed directly